### ISSUE TITLE
WordPress oEmbed support

### ISSUE THREAD

**Issue by spacedmonkey** [GoogleForCreators/web-stories-wp#1687]
## Feature Description

If you take the URL of a story and paste it into content of a WordPress site, a embedded preview appears. This preview should render a the web story correctly.

This is what the current preview looks like. 

<img width="823" alt="Screenshot 2020-05-08 at 16 19 45" src="https://user-images.githubusercontent.com/237508/81420366-cad6c600-9147-11ea-9ad4-8f2b337e6469.png">

When it should preview like this. 

<img width="453" alt="Screenshot 2020-05-08 at 16 21 03" src="https://user-images.githubusercontent.com/237508/81420503-f3f75680-9147-11ea-9967-2c2aaef846d1.png">

## Additional Context

Since WordPress embeds are sandboxed and have some restrictions, it's gonna be difficult to have a nice looking preview using amp-story-player. Needs some evaluation.

For history https://github.com/ampproject/amp-wp/issues/3072

Here's how WordPress embeds work for regular posts:

https://make.wordpress.org/core/2015/10/28/new-embeds-feature-in-wordpress-4-4/

----

**Comment by spacedmonkey** [comment]
Related: https://github.com/google/web-stories-wp/issues/1118

**Comment by swissspidy** [comment]
Since WordPress embeds are sandboxed and have some restrictions, it's gonna be difficult to have a nice looking preview like this.

**Comment by spacedmonkey** [comment]
I know that it might be hard, but we can style it a little better, IMO.

**Comment by swissspidy** [comment]
Agreed

**Comment by swissspidy** [comment]
For history https://github.com/ampproject/amp-wp/issues/3072

This needs a product spec before anything else.

Here's how WordPress embeds work for regular posts:

https://make.wordpress.org/core/2015/10/28/new-embeds-feature-in-wordpress-4-4/

**Comment by swissspidy** [comment]
This is becoming more relevant now that we have better ways for embedding stories. Would be good to have some improved UX here.

We can start with using the `Image` renderer for story oEmbeds.

In a next step/ticket we can look into making it dynamic (image with lightbox).

**Comment by spacedmonkey** [comment]
I guess that are two seniors we need to consider here. 

1. A user embeds a link to a story ( with web stories plugin installed )   
2. A user embeds a link to a story ( without web stories plugin installed )   

If a user does not have the plugin activate, we could use the image based embed with a link to the original article. This is likely to have wide browser / theme support. 

If a user embed a story from another site and does have the plugin, we could detect this and somehow, render the story. We could have to work out how the story was being rendered, do that other site have this plugin install or another plugin to build / render stories.

**Comment by swissspidy** [comment]
Let's do (1) for now (image based embed). Filtering the WP default oEmbed template parts in the plugin to be nicer when anyone tries to oEmbed it.

Some helpful resources for that:

https://make.wordpress.org/core/2015/10/28/new-embeds-feature-in-wordpress-4-4/
https://make.wordpress.org/core/2016/03/11/embeds-changes-in-wordpress-4-5/

This covers the case when someone else oEmbeds your story. They should look nice in that case.

> We could have to work out how the story was being rendered, do that other site have this plugin install or another plugin to build / render stories.

I don't see how the other site having the plugin installed would play a role in that case.

**Comment by spacedmonkey** [comment]
> I don't see how the other site having the plugin installed would play a role in that case.

With the web stories plugin installed, we know that web story can rendered on the front end. To do this, we need to enqueue scripts and styles. So with the plugin installed, we could find a way to render stories from a remote site as we do with "local" stories. Also if the plugin is installed / we check new the theme supports, we can have a reason exception that this embed will work without styling / js issues. 

The reason, at a guess that we would not to embed the full story on remote sites, as we have no idea of this embed will work correctly. Returning a simpler version of the story is the most likely to work in the most places.

**Comment by spacedmonkey** [comment]
I have a POC at [fix/oembed](https://github.com/google/web-stories-wp/tree/fix/oembed). 

### With Poster
![Screenshot 2021-03-29 at 15 49 04](https://user-images.githubusercontent.com/237508/112862671-e9221900-90ad-11eb-82c5-1f6769a632df.png)

### Without Poster. 
![Screenshot 2021-03-29 at 15 51 13](https://user-images.githubusercontent.com/237508/112862728-f6d79e80-90ad-11eb-9e6c-0c69626d8794.png)

Worthing nothing that Oembeds pass a requested height and width. I have used the requested width and ignored the height here. Instead used the width to generate a height with the correct aspect ratio. 


However, I found it hard to test locally as because of stories do no embed locally on the front end. To add any kind of E2E we will have to solve this issue first.
![Screenshot 2021-03-29 at 15 52 08](https://user-images.githubusercontent.com/237508/112862750-fc34e900-90ad-11eb-8b23-799cba0190e1.png)

**Comment by swissspidy** [comment]
We should not be filtering `embed_html`. WP should keep return the `<iframe>` code for embeds. We just make the iframe contents nicer via filtering the WP default oEmbed template parts in the plugin. The filter for that is `embed_template`, as described in detail in the blog posts linked above.

So replacing default iframe content (`wp-includes/theme-compat/embed-content.php`) with the result of `Image`.

> Without Poster.

Hmm we need a good alternative here. The output from `\Google\Web_Stories\Stories_Renderer\Renderer::render_single_story_content()` would be perfect here :/ But that needs refactoring in the future to make it reusable.

Perhaps we can just display the original embed content in that case for now?

> However, I found it hard to test locally as because of stories do no embed locally on the front end. To add any kind of E2E we will have to solve this issue first.

Does it work on a site without custom ports?

Worst case you could test it locally by adding the oEmbed HTML (the `<iframe>...` code) in a custom HTML block.

**Comment by spacedmonkey** [comment]
I have updated / fix the POC and now there is a PR at #6964 

The video shows well out it works. It works well as it is. I have not added tests, as I only managed to test it working locally by using VVV that does not use custom ports. This means until we fix the issue with custom ports, then E2E tests are out of the question ( unless we mock something in an e2e test plugin ? ). 

Not sure what test I can / should add to this?
